lib/marin/src/marin/transform/conversation/transform_conversation.py

```diff
@@ -332,17 +287,105 @@ def transform_hf_dataset(cfg: TransformSFTDatasetConfig):
             dataset = datasets.load_dataset(**dataset_kwargs)
             num_shards = dataset.num_shards
             if not num_shards:
-                raise ValueError(
-                    f"Streaming dataset {source} subset={subset} split={split} does not expose num_shards."
-                )
+                raise ValueError(f"Streaming dataset {source} subset={subset} split={split} does not expose num_shards.")
+
+            subset_name = subset or "default"
+            subset_output_path = get_shard_dir(cfg.output_path, subset_name, split)
+            output_path = create_shard_output_directory(subset_output_path)
 
+            # Yield a task for each shard
             for shard_idx in range(num_shards):
-                shard_refs.append(process_streaming_shard.remote(cfg, subset, split, shard_idx, num_shards))
+                yield ShardTask(
+                    source=source,
+                    revision=revision,
+                    subset=subset,
+                    split=split,
+                    shard_idx=shard_idx,
+                    num_shards=num_shards,
+                    output_path=output_path,
+                    cfg=cfg,
+                )
 
-    ray.get(shard_refs)
-    return cfg.output_path
+
+def process_shard_task(task: ShardTask) -> dict:
+    """Process a single shard of a dataset subset/split.
+
+    Loads a specific shard from HuggingFace Hub, transforms records, and writes to a single output file.
+    """
+    adapter = unwrap_versioned_value(task.cfg.adapter).copy()
+    if adapter is None:
+        raise ValueError("Transform configuration requires an adapter.")
+
+    dataset_kwargs: dict[str, object] = {
+        "path": task.source,
+        "split": task.split,
+        "streaming": True,
+        "revision": task.revision,
+    }
+    if task.subset not in (None, "default"):
+        dataset_kwargs["name"] = task.subset
+
+    dataset = datasets.load_dataset(**dataset_kwargs)
+    shard_dataset = dataset.shard(num_shards=task.num_shards, index=task.shard_idx)
+
+    subset_name = task.subset or "default"
+    output_filename = _shard_filename(task.output_path, task.shard_idx)
+    rows_written = 0
+
+    # Process and write shard
+    with fsspec.open(output_filename, "wt", compression="gzip") as f:
+        for raw_row in shard_dataset:
+            transformed_row = transform_row(raw_row, task.cfg, adapter)
+            if transformed_row is None:
+                continue
+
+            f.write(f"{json.dumps(transformed_row.model_dump())}\n")
+            rows_written += 1
+
```

Comment:
feels like this ought to fit into zephyr (also sorry i changed this file recently)